deviceModeInit.js

```diff
@@ -95,6 +96,117 @@ var rudderTracking = (function () {
       .catch((error) => {
         console.log("Error occurred while updating cart:", error);
       });
+
+    const productIsVisible = (window, element) => {
+      const viewportWidth = window.innerWidth;
+      const viewportHeight = window.innerHeight;
+      const { top, bottom, left, right, height } =
+        element.getBoundingClientRect();
+
+      if (
+        top < viewportHeight &&
+        bottom > 0 &&
+        left < viewportWidth &&
+        right > 0
+      ) {
+        let pixelsVisible = height;
+
+        if (top < 0) {
+          pixelsVisible += top;
+        }
+        if (bottom > viewportHeight) {
+          pixelsVisible += viewportHeight - bottom;
+        }
+        const percentVisible = pixelsVisible / height;
+
+        if (percentVisible > 0.8) {
+          return true;
+        }
+      }
+
+      return false;
+    };
+    function getAllProductsOnPage() {
+      const allAnchorTags = document.getElementsByTagName("a");
+      const productUrlRegex = new RegExp(
+        `^(?!\/\/cdn)[-\.:\/,a-z,A-Z,0-9]*\/products\/`
```

Comment:
this is very dependent on 'products' being present in the page url. there is also a support of collections, which are like all products of a brand under one place, do we not want to fire product list viewed event for those pages? As of now we are not doing so.

Example url: `store url`/collections/`collection-name`

Reply:
We are going to call this for all the pages. The regex is checking if the product URL which is assigned with the product image has \products\ substring or not.

Reply:
Yes, but it would not be firing for /collections/ only URLs, do we need to make a call in that case?

Reply:
We don't have to think about page-specific URLs as we will execute this on every page. Now if we found an anchor tag on the page with /products/ substring and that anchor tag also assigned to an image that means the use seen the product, then we will add that product to the product list viewed event

Reply:
@yashasvibajpai it looks for Anchor tags and it runs regex on it not the `page` url

Reply:
My bad, I had another thing in mind. Got your points.